### PR TITLE
chore: update flaky test for error reporting aggregation test

### PR DESCRIPTION
**Pull request by sanpj2292**
# Description

Flaky test for testing aggregation logic has been addressed through this PR

## Linear Ticket

Resolves INT-2232

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved JSON comparison in reporting tests to ensure accurate validation of data.
  - Updated error reporting logic to enhance the accuracy of error detail aggregation.

### CODE REVIEW

**Review by mihir20 — APPROVED** [review]
LGTM

**Review by ItsSudip — APPROVED** [review]

**Comment by lvrach** [comment]
Why does the JSON comparison fixes the problem? Could we use some other require function instead?

**Comment by achettyiitr** [comment]
> Why does the JSON comparison fixes the problem? Could we use some other require function instead?

I was wondering the same thing, we used to get something like [this](https://github.com/rudderlabs/rudder-server/actions/runs/9178103940/job/25237228883?pr=4687) kind of error when it fails.


```
    reporting_test.go:686: 
        	Error Trace:	/home/runner/work/rudder-server/rudder-server/enterprise/reporting/reporting_test.go:686
        	Error:      	Not equal: 
        	            	expected: []*types.EDMetric{(*types.EDMetric)(0xc000bb6240), (*types.EDMetric)(0xc000bb6300), (*types.EDMetric)(0xc000bb63c0)}
        	            	actual  : []*types.EDMetric{(*types.EDMetric)(0xc000bb6000), (*types.EDMetric)(0xc000bb60c0), (*types.EDMetric)(0xc000bb6180)}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -20,2 +20,9 @@
        	            	    (types.EDErrorDetails) {
        	            	+    StatusCode: (int) 400,
        	            	+    ErrorCode: (string) "",
        	            	+    ErrorMessage: (string) (len=32) "bad data sent for transformation",
        	            	+    EventType: (string) (len=8) "identify",
        	            	+    Count: (int64) 20
        	            	+   },
        	            	+   (types.EDErrorDetails) {
        	            	     StatusCode: (int) 200,
        	            	@@ -25,9 +32,2 @@
        	            	     Count: (int64) 10
        	            	-   },
        	            	-   (types.EDErrorDetails) {
        	            	-    StatusCode: (int) 400,
        	            	-    ErrorCode: (string) "",
        	            	-    ErrorMessage: (string) (len=32) "bad data sent for transformation",
        	            	-    EventType: (string) (len=8) "identify",
        	            	-    Count: (int64) 20
        	            	    }
        	Test:       	TestAggregationLogic
```

**Comment by sanpj2292** [comment]
I have used `JSONEq` in this case, as it equated JSON. I had assumed `[]object` also will be validated.
But the problem I have identified is in a specific field(Errors)'s ordering.
After fixing the ordering, there seems to be no flakiness actually. Can you guys verify @lvrach @achettyiitr ?

File: enterprise/reporting/error_reporting.go [hunk 1]
```diff
@@ -540,13 +540,14 @@ func (edr *ErrorDetailReporter) aggregate(reports []*types.EDReportsDB) []*types
...
-				StatusCode:   rep.StatusCode,
-				ErrorCode:    rep.ErrorCode,
-				ErrorMessage: rep.ErrorMessage,
-				EventType:    rep.EventType,
-				Count:        count,
+				StatusCode:   r.StatusCode,
+				ErrorCode:    r.ErrorCode,
+				ErrorMessage: r.ErrorMessage,
+				EventType:    r.EventType,
+				Count:        repCount,
```
[the remark this answers is not in the document]
Comment (sanpj2292) [reply enterprise/reporting/error_reporting.go]: The changes are done

**Review by achettyiitr — APPROVED** [review]

**Comment by achettyiitr** [comment]
> I have used `JSONEq` in this case, as it equated JSONs. I had assumed `[]object` also will be validated without considering ordering at all levels. But the problem I have identified is in a specific field(Errors)'s ordering. After fixing the ordering, there seems to be no flakiness actually. Can you guys verify @lvrach @achettyiitr ?

Yes, since there is no deterministic order for map, while iterating reportsCountMap would be giving output.

**Pull request merged**